ToDoApp/ToDoApp/ToDoServiceImpl.swift

```diff
@@ -0,0 +1,85 @@
+//
+//  ToDoServiceImpl.swift
+//  ToDoApp
+//
+//  Created by Trong Hieu Vuong on 2022-12-23.
+//
+
+import Foundation
+import FirebaseDatabase
+import FirebaseDatabaseSwift
+enum TaskError: Error {
+    case unavailableTask
+}
+
+struct TaskModel: Identifiable, Codable {
+    let id: UUID
+    var title: String
+    var deadline: Date
+    var status: Bool = false
+}
+
+struct ToDoTasks: Codable {
+    var today: [TaskModel]
+    var other: [TaskModel]
+}
+protocol DateChecker {
+    func isDateInToday(_ date: Date) -> Bool
+}
+//  conform Calendar to this protocol
+extension Calendar: DateChecker {}
+
+protocol ToDoService {
+    func createTask(title: String, deadline: Date) async -> Result<TaskModel, RepositoryError>
+    func updateTask(todo: TaskModel) async -> Result<TaskModel, RepositoryError>
+    func getTasks() async -> Result<ToDoTasks, RepositoryError>
+}
+
+final class ToDoServiceImpl: ToDoService {
+    // MARK: - Dependencies
+    private let dateChecker: DateChecker
+    private let repo: Repository
+    // MARK: - Init
+    init(dateChecker: DateChecker, repo: Repository) {
+        self.dateChecker = dateChecker
+        self.repo = repo
+    }
+    convenience init() {
+        self.init(dateChecker: Calendar.current, repo: Database.database().reference())
+    }
+    func createTask(title: String, deadline: Date) async -> Result<TaskModel, RepositoryError>{
+        let oneTask = TaskModel(id: UUID(), title: title, deadline: deadline)
+        let result = await repo.createOrUpdate(oneTask)
+        if (result != nil) {
+            return .failure(result!)
+        } else {
+            return .success(oneTask)
+        }
+    }
+    func updateTask(todo: TaskModel) async -> Result<TaskModel, RepositoryError> {
+        let result = await repo.createOrUpdate(todo)
+        if (result != nil) {
+            return .failure(result!)
```

Comment:
it is better to use unwrap syntax instead of force unwrapping
```suggestion
        let error = await repo.createOrUpdate(todo)
        if let error {
            return .failure(error)
```